Fix issue that loading order permutates per system

discordier opened the pull request:
As discussed on Mumble on 2020-07-02, we want reproducible builds (as everyone else).
However, Contao creates a `bundles.map` which can vary per host, directory, etc. despite using the very same `composer.lock` for input.
This fixes that behaviour by sorting the loading order keys and therefore have an deterministic output of the resolved dependencies.

discordier committed:
Fix issue that loading order permutates per system

As discussed on Mumble on 2020-07-02, we want reproducible builds (as everyone else).
However, Contao creates a `bundles.map` which can vary per host, directory, etc. despite using the very same `composer.lock` for input.
This fixes that behaviour by sorting the loading order keys and therefore have an deterministic output of the resolved dependencies.

ausi commented on src/Bundle/Config/ConfigResolver.php:
```diff
@@ -91,7 +91,8 @@ private function buildLoadingOrder(): array
             }
         }
-
+        ksort($loadingOrder);
```
```suggestion
        uksort($loadingOrder, function (string $a, string $b): int {
            return md5($a) <=> md5($b);
        });
```

discordier replied:
That is not deterministic anymore as `md5` might have collisions. :smile_cat:

leofeyer replied:
So now if I want to be first, I have to name my extension `j4qxp8` instead of `aaa_something`? Because the MD5 hash of `j4qxp8` is `05c28b9764f3e912c55b38fba3aafa97`. 🙈

discordier replied:
That would be the outcome I already mentioned on mumble and therefore I refrained from implementing this.
To be honest, my concern is that the output of the current system is unreliable and I try to fix this behaviour.

Making the life of developers hard via "You must not name your bundle so that you can predict the loading order via name" is beyond the scope of this PR.
This is already explicitely not supported as the loading order can only be noted via load after.
Loading a bundle explicitly before another one is not supported since Contao 4.0 pre discussions during Ahrweiler 1.
Implementing a hash like md5 or the like will shift the problem only instead of solving it.

My only concern was, that the current implementations output is unpredictable and I solved it.

If you really insist on making the order unpredictable, feel free to reverse the sorting order randomly in patch releases, as this is the most sufficient solution to this problem while maintaining reproducible results for a given version of Contao. :smirk:

ausi replied:
I think most developers that know how to get an md5 which starts with `0` while using a sane name understand that they shouldn’t do that. But we could easily prevent that with something like that:

```suggestion
        ksort($loadingOrder);
        $salt = md5(implode("\0", array_keys($loadingOrder)));

        uksort($loadingOrder, function (string $a, string $b) use($salt): int {
            return md5($a.$salt) <=> md5($b.$salt);
        });
```

However I think this is not necessary, so I would go with the simple sort by md5 hash from above.

fritzmg replied:
@ausi that would change the loading order when you install a new bundle, wouldn't it? That should be prevented imho, otherwise it's bad for the user experience in the back end again.

ausi replied:
> that would change the loading order when you install a new bundle, wouldn't it?

Not if we follow my last sentence in the comment 🙃 
> However I think this is not necessary, so I would go with the simple sort by md5 hash from above.

@aschempp Can you elaborate on the thumbs-down? Do you want to use the simple md5 version or do you want to sort it alphabetically, or something else?

fritzmg replied:
> Not if we follow my last sentence in the comment 🙃

Well yes, but then someone could influence the loading order by naming the bundle in a certain way (though I agree this is a negligible issue…).

ausi replied:
The simple `md5`ed sort has the benefit, that the same two bundles are always sorted the same way relative to each other, in every installation, everywhere. I think it is the perfect tradeof between stability/reproducibility and preventing bundle-name misuse.

discordier commented:
While working on this, I just discovered some more problems related to non deterministic states and even a bug handling of `loadAfter`.
I will update this PR and add the relevant tests.

Thanks to @ausi for converting it to draft and the fruitful discussion.

m-vo commented:
Just for the record, the issue we discussed was this:

Two bundles Foo + Bar both ship a dependency Lib and each create a bundle config for it:
```php
// Foo's Plugin#getBundles() returns
[
  BundleConfig::create(FooBundle::class),
  BundleConfig::create(LibBundle::class)->setLoadAfter([FooBundle::class])
]

// Bar's Plugin#getBundles() returns
[
  BundleConfig::create(BarBundle::class),
  BundleConfig::create(LibBundle::class)->setLoadAfter([BarBundle::class])
]
```

Is `Lib` now still guaranteed to be loaded after `Foo` and `Bar`? (i.e.: are the bundle config's constraints merged?)

Toflar commented:
Thanks for picking this up @discordier <3 [Reproducible builds](https://reproducible-builds.org/) FTW!

discordier committed:
Merge configurations when bundles are provided multiple times

When two extensions provide the same bundle to the system and both
provide a config for `loadAfter` and `replace`, the values must get
merged to a common array.
When at least one of each sets `loadInProduction` or
`loadInDevelopment`, the result must also get loaded then and not only
the one coming arbitrary "last".

Variance in config types is explicitely not handled so far and marked as
FIXME for the moment as I have no clue how to handle these and the use
case seems too narrow to handle them explicitely.

discordier committed:
Handle `setReplace` and  `setLoadAfter` sorted

To have predictable output, the results must be sorted, for
functionality the order does not matter so this is fine.

discordier commented:
I have updated my implementation and also added a [FIXME](https://github.com/contao/manager-plugin/pull/27/files#diff-be76487cdeb2e0d41287294fd3f8cbe7R67) we might want to discuss.

discordier committed:
Fix coding standard issues

discordier commented:
~~The coding standard issue in [here](https://github.com/contao/manager-plugin/pull/27/checks?check_run_id=834273991#step:5:40) is not trivial to resolve, as it causes test failure then.
However, it is correct that the order might change again, I will update this.
We can not switch to identity matching via same though, as the config itself will become a new, merged, config.
I refrained from using each of the original configs here, as they could, in theory, be used elsewhere and therefore manipulating them will open a new can of worms there then.~~

Implemented a foreach in dfb04e7

discordier committed:
Fix for assertEquals => assertSame in test

Now we have `assertSame` back at the expense of needing a loop.
However, the result is now guaranteed to exactly match the expected
output in order and contents.

discordier committed:
Yet another coding standard issue

ausi commented on tests/Bundle/Config/ConfigResolverTest.php:
```diff
@@ -51,13 +51,24 @@ public function testDefinesAFluentInterface(): void
+        // Shuffle the input around to ensure the input order does not alter the output.
+        // This does not guarantee failure if the random input already provides the expected output but is better
+        // than nothing - at least we have a higher probability than 0 to catch these issues here.
+        shuffle($configs);
```
```suggestion
        mt_srand(0);
        shuffle($configs);
        mt_srand();
```

Would make this test reproducible too. 
Or maybe run the test five times with different seeded random orders?

discordier replied:
Either sounds good to me.

ausi replied:
Running it five times with `mt_srand(0)` til `mt_srand(4)` would be my preferred solution.

discordier commented:
While playing with shuffling the bundle adding order, @baumannsven and I discovered that apparently even the [manager-bundle](https://github.com/contao/manager-bundle/blob/135667f2992daf20520df80ee20243e5e0f7e00a/src/ContaoManager/Plugin.php#L92) does not provide sufficient `loadAfter` definitions and depends on the adding order of bundles.
I have not yet tracked down WHY the loading order of extensions of plain symfony bundles suddenly matters when loaded via manager-bundle versus a plain symfony application but IMO the `loadAfter` rules must be added there as well or the behaviour of the extension loading must be adapted so that the order does NOT matter anymore.
Test case was:
1. swap `ksort($loadingOrder)` with `krsort($loadingOrder)`.
2. call `composer install` again.
2. suddenly `%kernel.secret%` (from FrameworkBundle) is not resolvable anymore.

I fear I'm up to open one can of worms after another here... did not think that such a small finding like the reproducable builds could generate such impact.

Currently we are shuffling the configs in the `Plugin` of the `contao-core` bundle.

discordier commented:
The results of our tests became contao/contao#1903, we were not able to find more issues.
I wonder if we now need to place a conflict here in the manager-plugin.

leofeyer commented:
What is the status of this issue?

discordier commented:
Waiting for decision(s) on:
- https://github.com/contao/manager-plugin/pull/27#discussion_r449656484 - Which of both should be implemented.
- can this [FIXME](https://github.com/contao/manager-plugin/pull/27/files#diff-be76487cdeb2e0d41287294fd3f8cbe7R67) somehow be improved or is the current implementation enough.
- As per https://github.com/contao/manager-plugin/pull/27#issuecomment-654457730 - do we need to add some conflicts?

discordier commented:
As discussed on Mumble on 2020-07-30:
- https://github.com/contao/manager-plugin/pull/27#discussion_r449656484
  - Running it five times with mt_srand(0) til mt_srand(4).
- can this [FIXME](https://github.com/contao/manager-plugin/pull/27/files#diff-be76487cdeb2e0d41287294fd3f8cbe7R67) somehow be improved or is the current implementation enough.
  - We keep the current implementation with the exception and revisit it again as soon as some one encounteres a problem.
- As per https://github.com/contao/manager-plugin/pull/27#issuecomment-654457730 - do we need to add some conflicts?
  - This PR will become part of a new minor version and add a conflict `"contao/manager-bundle": ">= 4.9.0 < 4.9.4"`

discordier committed:
Runtests five times with mt_srand(0) til mt_srand(4)

We now run the tests five times with different random seeds to ensure
the random sorting is predictable as well.
This is a compromise between having "random" input values and
predictable test output.

discordier committed:
Drop FIXME

After discussion we decided to keep the exception for the moment and
revisit this as soon as some one encounteres a problem with it.

discordier committed:
Add conflict with Contao <4.9.4

We now conflict with all previous 4.9 manager-bundle versions, as these
do not contain the fixes from contao/contao#1903 which are needed to not
break things for existing installations.

discordier committed:
Replace `ksort` with `md5` comparison

discordier commented:
All done from my side, ready for review! :tada:

Toflar reviewed (dismissed):
Neat!

leofeyer committed:
CI

leofeyer commented:
Thank you @discordier.
